### PR TITLE
raftstorev2: add basic layout

### LINKED ISSUE #12842: Support physical isolation between regions
RFC: tikv/rfcs#93

Summary:

Make every region stores its own data in an isolated LSM tree, in our case RocksDB.

This is the second step to make TiKV to operate fast and at low cost for big cluster.

The first step can be found in #11515.

Tasks:

- [ ] raftstore v2
    - [x] implement read
        - https://github.com/tikv/tikv/pull/13039
        - https://github.com/tikv/tikv/pull/13202
    - [x] implement write
    - [x] implement split
    - [ ] implement merge
    - [x] implement conf change
    - [x] coprocessor hooks
    - [x] memory trace
    - [x] flow control
    - [x] peer lifetime trace
    - [ ] use unified thread pool for apply batch system
    - [x] support driving basic raft activity
        - https://github.com/tikv/tikv/pull/12843
        - https://github.com/tikv/tikv/pull/12877
        - https://github.com/tikv/tikv/pull/12921
        - https://github.com/tikv/tikv/pull/13013
        - https://github.com/tikv/tikv/pull/13115
        - https://github.com/tikv/tikv/pull/13127
        - https://github.com/tikv/tikv/pull/13157
        - https://github.com/tikv/tikv/pull/13227
        - https://github.com/tikv/tikv/pull/13245
- [x] introduce tablet
- [x] implement snapshot

More to be added.

### PR DESCRIPTION
**Pull request by BusyJay**
### What is changed and how it works?
Issue Number: Ref #12842

What's Changed:

```commit-message
This is an attempt to reimplement raftstore using the new assumptions
that peer's range can be overlapped.

Currently, compatability is not considered, though we may think about
how to migrate from old version by the end of this year.

No concrete implementations is added yet, we may choose reuse
implementation from v1 or implementing new logic base on actual
requirement. The principle is
1. do not introduce history debt while reusing code as much as possible.
2. do not change the current implementations.
```

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by zhangjinpeng1987 — APPROVED** [review]
LGTM

File: components/raftstore-v2/src/router/message.rs [hunk 1]
```diff
@@ -0,0 +1,293 @@
+// Copyright 2016 TiKV Project Authors. Licensed under Apache-2.0.
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/router/message.rs:1]: 2022

File: components/raftstore-v2/src/raft/storage.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+};
+use slog::Logger;
+
+/// A storage for raft.
+pub struct Storage<ER> {
+    engine: ER,
+    logger: Logger,
+}
+
+impl<ER> Storage<ER> {
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/raft/storage.rs:16]: for Raft Storage,  what's the expecting difference between v2 and v1?

File: components/raftstore-v2/src/router/message.rs [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+            ResponseChannel::None => write!(fmt, "Callback::None"),
+            ResponseChannel::Read(_) => write!(fmt, "Callback::Read(..)"),
+            ResponseChannel::Write { .. } => write!(fmt, "Callback::Write(..)"),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[repr(u8)]
+pub enum PeerTick {
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/router/message.rs:137]: why not use V1's implementation as they're identical?

File: components/raftstore-v2/src/router/message.rs [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+            PeerTick::CheckLeaderLease,
+            PeerTick::ReactivateMemoryLock,
+            PeerTick::ReportBuckets,
+        ];
+        TICKS
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum StoreTick {
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/router/message.rs:187]: Same here. It seems identical to v1.

File: components/raftstore-v2/src/router/message.rs [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        RaftCommand {
+            request,
+            ch,
+            send_time: Instant::now(),
+        }
+    }
+}
+
+/// Message that can be sent to a peer.
+pub enum PeerMsg<EK: KvEngine> {
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/router/message.rs:228]: Comparing with V1, it does not have CasualMessage, HeartbeatPd, Destroy.   So are they not needed anymore in v2?

File: components/raftstore-v2/src/raft/storage.rs (hunk 2 above)
Comment (BusyJay) [reply components/raftstore-v2/src/raft/storage.rs:16]: The initialization is different.

File: components/raftstore-v2/src/router/message.rs (hunk 4 above)
Comment (BusyJay) [reply components/raftstore-v2/src/router/message.rs:187]: No, they are different. Several ticks are removed.

File: components/raftstore-v2/src/router/message.rs (hunk 3 above)
Comment (BusyJay) [reply components/raftstore-v2/src/router/message.rs:137]: Several ticks will be moved from `StoreTick` to `PeerTick`.

File: components/raftstore-v2/src/router/message.rs (hunk 5 above)
Comment (BusyJay) [reply components/raftstore-v2/src/router/message.rs:228]: No, I delete them on purpose and add them as needed in future development.

**Comment by tonyxuqqi** [comment]
Also regarding the assumptions:
1) peer's range can be overlapped: do you mean the actual tablet's data range or the region's range. 
2) "Currently, compatability is not considered, though we may think about
how to migrate from old version by the end of this year",  does it mean the TiKV won't work for both version until this end of the year? (Or do you mean we will essentially two version of tikv?)

**Comment by BusyJay** [comment]
>     1. peer's range can be overlapped: do you mean the actual tablet's data range or the region's range.

Both can be overlapped.
 
> does it mean the TiKV won't work for both version until this end of the year? (Or do you mean we will essentially two version of tikv?)

Yes, it will work with either one but not both at the same time. There will be only one version TiKV while can start with different version of raftstore.

**Review by tonyxuqqi — APPROVED** [review]

File: components/raftstore-v2/src/router/message.rs (hunk 1 above)
Comment (tisonkun) [reply components/raftstore-v2/src/router/message.rs:1]: ASF suggests using the interception year of the project consistently for all its files. But of course it's not a requirement.

* https://github.com/apache/skywalking/issues/10223
* https://issues.apache.org/jira/browse/LEGAL-629

**Pull request merged**